### ISSUE TITLE
Implement sandbox to run WAL redo routines in

### ISSUE THREAD

**Issue by hlinnaka** [zenithdb/zenith#36]

**Comment by kelvich** [comment]
One wild idea that came once during a talk with @ericseppanen is that if we will be able to compile and run postgres for a wasm target we also may use it as a sandbox to apply WAL and while wasm should be generally slower absence of inter-vm communication may speed things up. I don't expect that this approach would be really working for variety of reasons but it worth of investigating.

**Comment by hlinnaka** [comment]
Yeah, webassembly is worth investigating. Another interesting project I bumped into recently is https://gvisor.dev/. Also, there's https://github.com/rust-vmm for a full VM-level sandbox

**Comment by ericseppanen** [comment]
I threw out the wasm idea as a random crazy thought, but having thought about it more I don't think sandboxing is a great idea.

I would think that WAL processing is performance-critical. Having to cross a sandbox boundary seems like a lot of overhead.

If we could reimplement WAL handling in Rust, we would get a memory safety guarantee. That code would have to be free of Rust panics (which could be used as DoS attacks), but that's not too difficult.

There are some earlier comments above talking about bounds checks that I think are incorrect: Rust bounds checks are always enabled: there is no way (short of a buggy library using `unsafe` code) to disable bounds checks in release mode or in any other mode. The only things that I can think of that are off by default in release mode are integer overflow checks, and they can be turned back on in Cargo.toml.

**Comment by hlinnaka** [comment]
> If we could reimplement WAL handling in Rust, we would get a memory safety guarantee. That code would have to be free of Rust panics (which could be used as DoS attacks), but that's not too difficult.

Yes. But then we would have to reimplement the WAL handling in Rust. And keep maintaining that for every PostgreSQL major release. Sometimes there have been bug fixes in minor releases too. It's not a huge amount of code, but it's still a fair amount and it's pretty deep down and scattered in the Postgres sources. I would really rather not go there if we can find another way.

The first TODO here is to set up some performance testing, though, to see how much overhead we're really talking about. It might be a non-issue, given all the other overhead, or it could be super-critical. Or maybe it can be mitigated e.g. by writing a full-page image WAL record every 1000 modifications to a page in the compute node, so that there's less WAL to replay in the common cases.

On a more philosophical level, I wish we can find a good, fast, way to run safely Postgres code in page server and elsewhere. Aside from replaying WAL, that will open a lot of possibilities like doing logical decoding, and running VACUUM and other background jobs in the background.

**Comment by ericseppanen** [comment]
I chatted with @kelvich about this. I will try to summarize:
- wasm will take an unknown amount of effort to work. I'm scared of this becoming an infinite time-sink.
- It's probably too early to talk about pushing this work to a cloud-level VM.
- Firecracker VMs look OK. Should investigate.
- containers don't provide great security guarantees (large syscall surface area). dockerd is overkill, runc is a little under-developed and hard to get right.
- Linux seccomp looks OK. Should investigate.

We didn't talk much about the "native Rust" option. My current thinking:

"Rewrite it in rust" may take a lot of time, and will need constant maintenance to stay in sync with new Postgres releases. So even if we might like this as a long-term option, it would be nice to have something easy in the short term. Plus, even if we do a Rust wal_redo implementation, it would be good to keep the C wal_redo_main so we can verify that the two implementations return the same result.

**Comment by hlinnaka** [comment]
https://github.com/zenithdb/postgres/pull/52 will take care of this, with seccomp

**Comment by mklgallegos** [comment]
hi @funbringer  it looks like this issue is waiting on a code review from you. Not trying to bother, just wanted to check in since this issue is schedule to be completed in milestone v0.1 (concludes at end of the month).  

Let me know if this looks like something that will slip past the end of the month.

**Comment by funbringer** [comment]
@mklgallegos this patch has been superseded by https://github.com/zenithdb/postgres/pull/56, and I've already pushed a bunch of bug-fixes in https://github.com/zenithdb/postgres/pull/58. There's also an on-going discussion [here](https://github.com/zenithdb/postgres/commit/8478beabaf7d6916d22bde99ee5a577a9b6b237c#commitcomment-53448914), which will hopefully be resolved after I've finished yet another patch.

**Comment by mklgallegos** [comment]
Hi @funbringer - this item was put into the done category during the milestone v0.2 planning session on 2021-07-30. I'm closing out. Please re-open if I am mistaken.
